### PR TITLE
feat(api): Adds unstable_pages module to JS API

### PR DESCRIPTION
**Pull request by jrf0110**
What this PR solves / how to test:

This PR refactors the existing Pages publish code into its own exportable function and exports it in the wrangler API under the `unstable_pages` object. This will allow folks to run the `publish` function in a scriptable JS environment and will get the entire deployment response as a variable.

Associated docs issues/PR:

- https://github.com/cloudflare/wrangler2/issues/2437

Author has included the following, where applicable:

- [ ] ~~Tests~~
- [x] Changeset

Reviewer has performed the following, where applicable:

- [x] Checked for inclusion of relevant tests
- [x] Checked for inclusion of a relevant changeset
- [x] Checked for creation of associated docs updates
- [x] Manually pulled down the changes and spot-tested

### CODE REVIEW

**Comment by JacobMGEvans** [comment]
I will pull this down and test it tomorrow, unless someone else gets to it first. At a quick glance it seems like just turning the current code programmatic and moving it under the `/api` which as long as the tests are passing should be a confident refactor.

File: packages/wrangler/src/api/pages/publish.tsx
Comment (jrf0110) [inline packages/wrangler/src/api/pages/publish.tsx:122]: I've added this new `customFunctionsDirectory` property that allows consumers to pass in a path to the functions directory, instead of relying on current working directory (like the CLI does)

File: packages/wrangler/src/api/pages/publish.tsx
Comment (GregBrimble) [inline packages/wrangler/src/api/pages/publish.tsx:60]: Since this is unstable, I'm happy for this to just be done in a follow-up if you can work around it in your client right now. If you _want_ to do it, go ahead, but it's not blocking.

File: packages/wrangler/src/api/pages/publish.tsx
Comment (GregBrimble) [reply packages/wrangler/src/api/pages/publish.tsx:123]: I have thoughts and feelings on this for when we think about stabilising this API, but totally fine right now since it's unstable.

**Review by GregBrimble — APPROVED** [review]
LGTM. But would be good to get someone from Wrangler proper to sign off on us adding this API before merging.

File: packages/wrangler/src/__tests__/pages.test.ts
Comment (jrf0110) [inline packages/wrangler/src/__tests__/pages.test.ts:2569]: @GregBrimble I found it more reliable to not sort the requests and lookup by index. But rather key the requests by the `key` property and lookup by said `key` property

File: packages/wrangler/src/api/index.ts
Comment (rozenmd) [inline packages/wrangler/src/api/index.ts:3]: You might want to export the types for `unstable_pages` too?

Comment (jrf0110) [reply packages/wrangler/src/api/index.ts:3]: If you don't mind, I'd like to hold off on that rabbit hole. There's a lot of types involved and I'd like to move more code around to export all of the relevant bits. If a consumer needs the type of the argument to `publish` or the return type, then they can easily infer it:

```tsx
type PublishOptions = typeof unstable_pages.publish extends (options: infer P) => any ? P : never
type PublishResult = typeof unstable_pages.publish extends (options: any) => infer P ? P : never
```

I - being the only consumer of this code so far - don't actually need that yet :)

**Pull request merged**
